cmd/builder/internal/builder/main.go

```diff
@@ -230,6 +242,64 @@ func (c *Config) allComponents() []Module {
 						*c.Providers...)...)...)...)...)
 }
 
+func (c *Config) updateModules() error {
+	if !c.SkipNewGoModule {
+		return nil
+	}
+
+	// Build the main service dependency
+	coremod, corever := c.coreModuleAndVersion()
+	corespec := coremod + " " + corever
+
+	if err := c.updateGoModule(corespec); err != nil {
+		return err
+	}
+
+	for _, comp := range c.allComponents() {
+		if err := c.updateGoModule(comp.GoMod); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *Config) updateGoModule(modspec string) error {
+	// Re-parse the go.mod file on each iteration, since it can
+	// change each time.
+	modulePath, dependencyVersions, err := c.readGoModFile()
+	if err != nil {
+		return err
+	}
+
+	mod, ver, _ := strings.Cut(modspec, " ")
+	if mod == modulePath {
+		// this component is part of the same module, nothing to update.
+		return nil
+	}
+
+	// check for exact match
+	hasVer, ok := dependencyVersions[mod]
+	if ok && hasVer == ver {
+		c.Logger.Info("Component version match", zap.String("module", mod), zap.String("version", ver))
+		return nil
+	}
+
+	scomp := semver.Compare(hasVer, ver)
+	if scomp > 0 {
+		// version in enclosing module is newer, do not change
+		c.Logger.Info("Not upgrading component, enclosing module is newer.", zap.String("module", mod), zap.String("existing", hasVer), zap.String("config_version", ver))
+		return nil
+	}
+
+	// upgrading or changing version
+	updatespec := mod + "@" + ver
+
+	if _, err := runGoCommand(*c, "get", updatespec); err != nil {
```

Comment:
I am not sure which one would be better, but we also have the option of doing `go mod edit -require=example.org@v1.2.0`. Why is `go get` preferred here?

Reply:
From what I can tell, the difference is that `go get` installs and builds the packages needed whereas `go mod edit` doesn't.  Given that we run `go mod tidy` after this in `GetModules()`, I'll change this to use `go mod edit`.

Reply:
One thing to [note](https://go.dev/ref/mod#go-mod-edit) about using `go mod edit -require`:
>These flags are mainly for tools that understand the module graph. Users should prefer go get path@version or go get path@none, which make other go.mod adjustments as needed to satisfy constraints imposed by other modules. 

I'm not sure what these other adjustments are or if we need them here? I would expect if we change the go.mod into some incompatible graph that `go mod tidy` will fail after this.

Reply:
Thinking more about this, I think I would prefer using `go mod edit -require` at first. The section you quoted sounds to me as "`go get` also does `go mod tidy`, which you probably want", and since, as you point out, we already `go mod tidy` after this, I would prefer to keep things simple